File: databuild/setup_transunion.R
# Load data ####

tu_data <- list.files(tu_data_dir, recursive=TRUE, full.names=TRUE) %>% 
  str_subset("CSV") %>%
  read_delim(delim="|") %>%
  bind_rows

tu_demographic <- list.files(tu_data_dir, recursive=TRUE, full.names=TRUE) %>% 
  str_subset("PSV") %>%
  read_delim(delim="|") %>%
  bind_rows

tu_data <- tu_data %>%
  left_join(tu_demographic,
            by=c("customerInput_tusequencenumber"="IN1"))

tu_data <- tu_data %>%
  transmute(state=customerInput_state,
            zip=customerInput_zip,
            state_mail=customerInput_state1,
            zip_mail=customerInput_zip1,
            tu_id=customerInput_tusequencenumber %>% as.numeric(),
            tu_permid=customerInput_customerInputPermId,
            credit_date=creditAsOfDate_creditAsOfDate,
            credit_score=cvtg03_finscore %>% as.numeric(),
            advrs1=cvtg03_advrs1 %>% as.numeric(),
            advrs2=cvtg03_advrs2 %>% as.numeric(),
            advrs3=cvtg03_advrs3 %>% as.numeric(),
            advrs4=cvtg03_advrs4 %>% as.numeric(),
            edie=edie04_score %>% as.numeric(),
            etie=etie04_score %>% as.numeric(),
            hh_income_estimate=EstHHIncome %>% as.numeric(),
            ethnicity=case_when(
              !is.na(GroupD) ~ 1,
              !is.na(GroupB) ~ 0,
              .default=NA)) %>%
  distinct()


# Save ####
save(tu_data,
     file=paste0(working_data_dir, "/tu_data.RData"))

File: databuild/transunion_queries/setup_transunion.py
# Load data ####
tu_data_files = glob.glob(os.path.join(raw_tu_data_dir, '**', '*CSV*'), recursive=True)
tu_data = pd.concat([pd.read_csv(f, delimiter='|', dtype=str) for f in tu_data_files], ignore_index=True)

tu_demographic_files = glob.glob(os.path.join(raw_tu_data_dir, '**', '*PSV*'), recursive=True)
tu_demographic = pd.concat([pd.read_csv(f, delimiter='|', dtype=str) for f in tu_demographic_files], ignore_index=True)

# Merge
tu_data = tu_data.merge(tu_demographic, left_on='customerInput_tusequencenumber', right_on='IN1', how='left')

# Transform
def to_numeric(s):
    return pd.to_numeric(s, errors='coerce')

tu_data = pd.DataFrame({
    'state': tu_data['customerInput_state'],
    'zip': tu_data['customerInput_zip'],
    'state_mail': tu_data['customerInput_state1'],
    'zip_mail': tu_data['customerInput_zip1'],
    'tu_id': to_numeric(tu_data['customerInput_tusequencenumber']),
    'tu_permid': tu_data['customerInput_customerInputPermId'],
    'credit_date': tu_data['creditAsOfDate_creditAsOfDate'],
    'credit_score': to_numeric(tu_data['cvtg03_finscore']),
    'advrs1': to_numeric(tu_data['cvtg03_advrs1']),
    'advrs2': to_numeric(tu_data['cvtg03_advrs2']),
    'advrs3': to_numeric(tu_data['cvtg03_advrs3']),
    'advrs4': to_numeric(tu_data['cvtg03_advrs4']),
    'edie': to_numeric(tu_data['edie04_score']),
    'etie': to_numeric(tu_data['etie04_score']),
    'hh_income_estimate': to_numeric(tu_data['EstHHIncome']),
    'ethnicity': tu_data.apply(lambda row: 1 if pd.notna(row['GroupD']) else (0 if pd.notna(row['GroupB']) else pd.NA), axis=1)
}).drop_duplicates()

# Save as .parquet (pyarrow format, R-compatible)
tu_data.to_parquet(os.path.join(working_data_dir, 'pre-processed', 'tu_data.parquet'), index=False)